Continue publishing md5s / sha1s, put everything into shaNNNsums

justinsb opened the pull request:
This script appears to be used for previous versions, at least in CI,
so we should keep publishing sha1 & md5 files.

Also put all our assets into the SHA256SUMS / SHA512SUMS files.  This
means that we'll end up signing everything we publish (at little to no
cost).  It also means that SHA256SUMS and SHA512SUMS files can be used
for mirroring.  Also, we can detect _extra_ files, which is also
potentially a problem for release attestation.  The implementation
runs the hash twice, but the code is much simpler for it.

We also unroll the loop for clarity, and so we can easily
version-guard md5/sha1 as part of a future deprecation.

justinsb commented:
/assign @justaugustus

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +680,13 @@ release::gcs::locally_stage_release_artifacts() {
-      cat "$sha_file" >> "${gcs_stage}/SHA${bits}SUMS"
-    done
+    common::sha $path 256 "full" >> "${gcs_stage}/SHA256SUMS" || return 1
+    common::sha $path 512 "full" > "${gcs_stage}/SHA512SUMS" || return 1
```
```suggestion
    common::sha $path 512 "full" >> "${gcs_stage}/SHA512SUMS" || return 1
```

tpepper commented on lib/releaselib.sh:
```diff
@@ -680,17 +680,13 @@ release::gcs::locally_stage_release_artifacts() {
-    done
-  done
+    common::md5 $path > "$path.md5" || return 1
+    common::sha $path 1 > "$path.sha1" || return 1
```
I suggest we condition these two lines to run only if RELEASE_BRANCH is release-1.13, release-1.14, or release 1.15 and open an issue to remove this small block in six months when 1.15 goes out of support.

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
-    done
-  done
+    if [[ -n "${publish_old_hashes}" ]]; then
+      common::md5 $path > "$path.md5" || return 1
```
```suggestion
      common::md5 "$path" > "${path}.md5" || return 1
```

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
-  done
+    if [[ -n "${publish_old_hashes}" ]]; then
+      common::md5 $path > "$path.md5" || return 1
+      common::sha $path 1 > "$path.sha1" || return 1
```
```suggestion
      common::sha "$path" 1 > "${path}.sha1" || return 1
```

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
-    for sha_file in "${gcs_stage}"/*.sha"${bits}"; do
-      cat "$sha_file" >> "${gcs_stage}/SHA${bits}SUMS"
-    done
+    common::sha $path 256 > "$path.sha256" || return 1
```
```suggestion
    common::sha "$path" 256 > "${path}.sha256" || return 1
```

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
-      cat "$sha_file" >> "${gcs_stage}/SHA${bits}SUMS"
-    done
+    common::sha $path 256 > "$path.sha256" || return 1
+    common::sha $path 512 > "$path.sha512" || return 1
```
```suggestion
    common::sha "$path" 512 > "${path}.sha512" || return 1
```

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
+    common::sha $path 256 > "$path.sha256" || return 1
+    common::sha $path 512 > "$path.sha512" || return 1
+
+    common::sha $path 256 "full" >> "${gcs_stage}/SHA256SUMS" || return 1
```
```suggestion
    common::sha "$path" 256 "full" >> "${gcs_stage}/SHA256SUMS" || return 1
```

hoegaarden commented on lib/releaselib.sh:
```diff
@@ -680,17 +698,16 @@ release::gcs::locally_stage_release_artifacts() {
+    common::sha $path 512 > "$path.sha512" || return 1
+
+    common::sha $path 256 "full" >> "${gcs_stage}/SHA256SUMS" || return 1
+    common::sha $path 512 "full" >> "${gcs_stage}/SHA512SUMS" || return 1
```
```suggestion
    common::sha "$path" 512 "full" >> "${gcs_stage}/SHA512SUMS" || return 1
```

hoegaarden commented on lib/common.sh:
```diff
@@ -638,6 +638,20 @@ common::argc_validate () {
+  local file=$1
+
+  if which md5 >/dev/null 2>&1; then
+    md5 -q "$1"
```
nit: is there a reason, why we don't use `$file` here?
```suggestion
    md5 -q "$file"
```

justinsb commented:
* Added quotes around args (thanks @hoegaarden )
* Added test vectors for md5 & sha1
* Added comment & link to test run explaining how kops detected that this was going to affect 1.14 and 1.15 releases :-)

justaugustus commented:
([Running a mock stage build](https://console.cloud.google.com/gcr/builds/9fffb762-4ba9-472d-844f-1cc703b34cce?project=648026197307) to validate the outputs from this PR.)

justaugustus commented:
/lgtm
/approve
/hold cancel

BenTheElder commented:
https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-kops-aws/1167500303809908737

> W0830 18:14:12.258] I0830 18:14:12.254653     405 builder.go:297] error reading hash file "https://storage.googleapis.com/kubernetes-release-dev/ci/v1.17.0-alpha.0.1022+e176e477195f92/bin/linux/amd64/kubelet.sha1": file does not exist
